crucible-llvm/src/Lang/Crucible/LLVM/Ctors.hs

```diff
@@ -0,0 +1,178 @@
+import           Lang.Crucible.CFG.Reg (Expr(App))
+import qualified Lang.Crucible.CFG.Reg as Reg
+import           Lang.Crucible.CFG.SSAConversion (toSSA)
+import           Lang.Crucible.FunctionHandle (HandleAllocator, mkHandle')
+import           Lang.Crucible.Types (UnitType, TypeRepr(UnitRepr))
+import           Lang.Crucible.LLVM.Extension (LLVM, ArchWidth)
+import           Lang.Crucible.LLVM.Intrinsics (LLVMContext, _llvmTypeCtx)
+import           Lang.Crucible.LLVM.Types (HasPtrWidth)
+
+{- Example:
+
+@llvm.global_ctors = appending global [3 x { i32, void ()*, i8* }] [{ i32, void ()*, i8* } { i32 65535, void ()* @_GLOBAL__sub_I_HkdfTest.cpp, i8* null }, { i32, void ()*, i8* } { i32 65535, void ()* @_GLOBAL__sub_I_gtest_all.cc, i8* null }, { i32, void ()*, i8* } { i32 65535, void ()* @_GLOBAL__sub_I_iostream.cpp, i8* null }]
+
+-}
+
+-- | A representation of well-typed inhabitants of the @llvm.global_ctors@ array
+--
+-- See https://llvm.org/docs/LangRef.html#the-llvm-global-ctors-global-variable
+data Ctor = Ctor
+  { ctorPriority :: Integer
+  , ctorFunction :: L.Symbol
+  , ctorData     :: Maybe L.Symbol
+  } deriving (Data, Eq, Generic, Ord, Show, Typeable)
+
+-- | Get the global variable representing @llvm.global_ctors@.
+getGlobalCtorsGlobal :: L.Module -> Maybe L.Global
+getGlobalCtorsGlobal mod_ =
+  let symb = L.Symbol "llvm.global_ctors"
+  in find (\x -> L.globalSym x == symb) (L.modGlobals mod_)
+
+-- | Unpack a @ctors@ value of type @{ i32, void ()*, i8* }@ from the AST
+extractCtors :: L.Value -> Maybe Ctor
+extractCtors val =
+  case val of
+    -- This is permissive about the integer widths... No reason to get caught up.
+    L.ValStruct [ L.Typed (L.PrimType (L.Integer _w0)) (L.ValInteger priority)
+                , L.Typed (L.PtrTo (L.FunTy (L.PrimType L.Void) [] _bool)) (L.ValSymbol symb)
+                , L.Typed (L.PtrTo (L.PrimType (L.Integer _w1))) data0_
+                ] -> Just . Ctor priority symb $
+                       case data0_ of
+                         L.ValSymbol data_ -> Just data_
+                         _                 -> Nothing
+    _ -> Nothing
+
+-- | Unpack and sort the values in @llvm.global_ctors@ by priority
+globalCtors :: (MonadError String m)
+            => L.Module
+            -> m [Ctor]
+globalCtors mod_ =
+  case getGlobalCtorsGlobal mod_ >>= L.globalValue of -- in the Maybe monad
+    Just (L.ValArray _ty vs) -> do
+
+      -- Assert that each value is of the expected type.
+      vs' <- forM vs $ \v ->
+        fromMaybe
+          (throwError $ unlines $ [ "Ill-typed value in llvm.global_ctors: "
+                                  , show v
+                                  ])
+          (pure <$> extractCtors v)
+
+      -- Sort the values by priority, highest to lowest.
+      pure (sortBy (comparing (Down . ctorPriority)) vs')
+
+    Nothing -> throwError "Couldn't find global llvm.global_ctors"
```

Comment:
Is this an error condition? If there aren't any ctors, does clang emit this variable at all?

Reply:
Good question. The motivation here is that a user can toggle whether this function is getting called in saw-script, and it won't be by default. Thus, if the user enables `global_ctors` and they aren't present, that seems like a user error. 

We could change this so that the caller passes in an `L.Global`, but we would have to catch the case where they pass in the wrong one... What do you think?